src/components/QrcodeStream.vue

```diff
@@ -55,31 +55,68 @@ import { assert } from '../misc/util'
 
 const props = defineProps({
   // in this file: don't use `props.constraints` directly. Use `constraintsCached`.
+  /**
+   * Passes an object with various camera configuration options.
+   */
   constraints: {
     type: Object as PropType<MediaTrackConstraints>,
     default() {
       return { facingMode: 'environment' } as MediaTrackConstraints
     }
   },
   // in this file: don't use `props.formats` directly. Use `formatsCached`.
+  /**
+   * Passes formats that will be recognized during detection.
+   */
   formats: {
     type: Array as PropType<BarcodeFormat[]>,
     default: () => ['qr_code'] as BarcodeFormat[]
   },
+  /**
+   * Setting this prop to true freezes the camera. Set to false to resume.
+   */
   paused: {
     type: Boolean,
     default: false
   },
+  /**
+   * Enables or disables camera torch during detection.
+   */
   torch: {
     type: Boolean,
     default: false
   },
+  /**
+   * Defines callback function that will be responsible for drawing detected code tracking rectangle
+   */
   track: {
-    type: Function
+    type: Function as PropType<
+      (detectedCodes: AdjustedCode[], ctx: CanvasRenderingContext2D | null) => void
+    >
   }
 })
 
-const emit = defineEmits(['detect', 'camera-on', 'camera-off', 'error'])
+const emit = defineEmits<{
+  /**
+   * Defines callback function called when code detetect.
+   */
+  (e: 'detect', detectedCodes: DetectedBarcode[]): void
+  /**
+   * Defines callback function called when camera becomes on.
+   */
+  (e: 'camera-on', capabilities: Partial<MediaTrackCapabilities>): void
+  /**
+   * Defines callback function called when camera becomes off.
+   */
+  (e: 'camera-off'): void
+  // TODO: `error` should have more precise typing.
+  //        According to https://gruhn.github.io/vue-qrcode-reader/demos/Simple.html example returned error seems to have known structure
+  //        Is it ok to write interface that statisfies whats in that example?
+  /**
```

Comment:
Good point. You can probably also import the error types from here: https://github.com/gruhn/vue-qrcode-reader/blob/master/src/misc/errors.ts

Reply:
(the remaining errors should be built in error types)

Reply:
So in general `Error` type should be returned maybe with union to https://github.com/gruhn/vue-qrcode-reader/blob/master/src/misc/errors.ts imported types?

Reply:
I have pushed the changes

Reply:
> So in general Error type should be returned maybe with union to https://github.com/gruhn/vue-qrcode-reader/blob/master/src/misc/errors.ts imported types?

Ah right, just `Error` is probably also fine. I think a union like `SpecialError | Error` is just the same as `Error`. All those custom error types are sub-types of `Error` and a union of a type with it's super-type is just the super-type again. Similar to how `"foo" | string` is just `string`.

Reply:
Should I create new PR with next commit or you can pull latest changes?

Reply:
Not sure what you mean. I would suggest to turn that error type into just `Error` and remove those `TODO` comments (if you consider those done) and then we can merge I would say. What do you think?

Reply:
Right! I have forgotten to remove the comments.

Please pull latest changes from `true-vue:emit+props-types_improved`

Reply:
Thanks :+1: 

What about the error type though? My point was that this definition:
```ts
type EmmitedError = DropImageFetchError | StreamApiNotSupportedError | InsecureContextError | StreamLoadTimeoutError | Error
```
is exactly the same as just: 
```ts
type EmmitedError = Error
```
[Checkout this TypeScript playground link.](https://www.typescriptlang.org/play/?&q=105#code/MYGwhgzhAEDKAOBTYBLMICiAnLB7L0iAHgC6IB2AJjNngQN4BQ00wu5EJWArsCfgAoAlNHrQI3JFgEByBMjSYc+GSIC+jZuNwBbRBlJYwAWUQ6ARomkimLFlkQluWctAAsAJi0aNjEgE8kaAAVRE4ARmgAXmgBWnxoAB84JFR0eKwRYjIqGmUCAH5oLm5EaAAuaAAzdAhERgB6BrtoAD0CzQCg0M4PaOgMwlIKaljB5Pk0pToRIpKyypqQOsbmu3agA). And therefore there is no need to define `EmmitedError`. We can just use `Error` directly.

Reply:
I understand your way of thinking and reasoning: for type resolving `Error` is just enough. Event with type narrowing it should work fine:

```
function handleEmittedError(err: Error) {
  if (err instanceof DropImageFetchError) {
    // specific handling
  } else if (err instanceof StreamApiNotSupportedError) {
    // different handling
  } else {
    // fallback or default error handling
  }
}
```

But for the sake of explicitly and documenting I would sustain union type in this case. It will inform code consumer that "under the hood" more specific error types might be available.

```
// EmmitedError type might give consumer additional clues: 
// Ok, that might be something more than standard error I'll check inside... and that would allow him or her to explore handling possiblities.
function handleEmittedError(err: EmmitedError) {
  ...
}
```

Thats my opinion on organizing the code - you are the product owner you decide. I can "collapse" the to simple `Error` if you decide. Just give me your decision :)

Reply:
Ok fair enough 👍